src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp

```diff
@@ -1342,47 +1342,26 @@ CXformUtils::PexprLogicalDMLOverProject(CMemoryPool *mp,
 	CColRef *pcrAction = NULL;
 	CColRef *pcrOid = NULL;
 
-	if (ptabdesc->IsPartitioned())
+	CExpressionArray *pdrgpexprProjected = GPOS_NEW(mp) CExpressionArray(mp);
+	// generate one project node with two new columns: action, oid (based on the traceflag)
+	pdrgpexprProjected->Append(CUtils::PexprScalarConstInt4(mp, val));
+
+	BOOL fGeneratePartOid = CUtils::FGeneratePartOid(ptabdesc->MDId());
+	if (fGeneratePartOid)
 	{
-		// generate a PartitionSelector node which generates OIDs, then add a project
-		// on top of that to add the action column
-		CExpression *pexprSelector = PexprLogicalPartitionSelector(
-			mp, ptabdesc, colref_array, pexprChild);
-		if (CUtils::FGeneratePartOid(ptabdesc->MDId()))
-		{
-			pcrOid = CLogicalPartitionSelector::PopConvert(pexprSelector->Pop())
-						 ->PcrOid();
-		}
-		pexprProject = CUtils::PexprAddProjection(
-			mp, pexprSelector, CUtils::PexprScalarConstInt4(mp, val));
-		CExpression *pexprPrL = (*pexprProject)[1];
-		pcrAction = CUtils::PcrFromProjElem((*pexprPrL)[0]);
```

Comment:
Perhaps, for HAWQ and possible implementation of GPDB this branch with adding of PartitionSelector node emitting partition oid value for Insert node makes sense. I would leave it and remove PartitionSelector just in not `CUtils::FGeneratePartOid(ptabdesc->MDId())` case

Reply:
Proposed change instead your one:
```C
diff --git a/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp b/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp
index 04ca5e26cb..cbfdd2618f 100644
--- a/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp
+++ b/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp
@@ -1344,17 +1344,21 @@ CXformUtils::PexprLogicalDMLOverProject(CMemoryPool *mp,
 
     if (ptabdesc->IsPartitioned())
     {
-        // generate a PartitionSelector node which generates OIDs, then add a project
-        // on top of that to add the action column
-        CExpression *pexprSelector = PexprLogicalPartitionSelector(
-            mp, ptabdesc, colref_array, pexprChild);
         if (CUtils::FGeneratePartOid(ptabdesc->MDId()))
         {
+            // generate a PartitionSelector node which generates OIDs, then add a project
+            // on top of that to add the action column
+            CExpression *pexprSelector = PexprLogicalPartitionSelector(
+                mp, ptabdesc, colref_array, pexprChild);
+
             pcrOid = CLogicalPartitionSelector::PopConvert(pexprSelector->Pop())
                          ->PcrOid();
+            pexprProject = CUtils::PexprAddProjection(
+                mp, pexprSelector, CUtils::PexprScalarConstInt4(mp, val));
+        } else {
+            pexprProject = CUtils::PexprAddProjection(
+                mp, pexprChild, CUtils::PexprScalarConstInt4(mp, val));
         }
-        pexprProject = CUtils::PexprAddProjection(
-            mp, pexprSelector, CUtils::PexprScalarConstInt4(mp, val));
         CExpression *pexprPrL = (*pexprProject)[1];
         pcrAction = CUtils::PcrFromProjElem((*pexprPrL)[0]);
     }
```

Reply:
Thanks, I'll add them to the PR

Reply:
I meant this change as replace of all your one.